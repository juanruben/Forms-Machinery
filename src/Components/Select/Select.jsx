import React from 'react';
import PropTypes from 'prop-types';
import './Select.scss';

const Select = (props) => {
    const {
        label, options, errors, name, value, onChange, placeholder, readOnly, hideReadOnly, required,
    } = props;
    const warning = (errors[name] && errors[name].length > 0);

    if (hideReadOnly && readOnly) {
        return null;
    }

    if (readOnly) {
        const selected = options.find((item) => item.id === value);
        return (
            <div className="read-only-container">
                <div className="label">{label}</div>
                <div className="read-only-value">{selected ? selected.name : ''}</div>
            </div>
        );
    }

    return (
        <>
            <div className="select-container">
                {label}
                {required ? ' *' : ''}
                <select className={`${warning && 'border-error'}`} name={name} id="" value={value} onChange={onChange}>
                    <option value="0" disabled>{placeholder}</option>
                    {options.map((item) => (
                        <option key={item.id} value={item.id}>{item.name}</option>
                    ))}
                </select>
                {warning && <div className="warning">{errors[name]}</div>}
            </div>
        </>
    );
};

Select.propTypes = {
    label: PropTypes.string,
    errors: PropTypes.object,
    options: PropTypes.array,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string,
    placeholder: PropTypes.string,
    readOnly: PropTypes.bool,
    hideReadOnly: PropTypes.bool,
    required: PropTypes.bool,
};

Select.defaultProps = {
    label: '',
    errors: {},
    options: [],
    value: '0',
    placeholder: '',
    readOnly: false,
    hideReadOnly: false,
    required: false,
};

export default Select;
