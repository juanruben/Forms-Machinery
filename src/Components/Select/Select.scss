@import "../../styles.scss";

.read-only-container {
  .label {
    font-weight: bold;
  }

  .read-only-value {
    padding-bottom: 20px;
    padding-left: 5px;
  }
}

.select-container {
  position: relative;

  select {
    border: solid $icafal-orange 1px;
    border-radius: 3px;
    display: block;
    height: 38px;
    margin-bottom: 10px;
    padding-left: 10px;
    width: 100%;
  }

  .border-error {
    border: solid red 3px;
    border-radius: 4px;
  }

  .warning {
    color: red;
    position: absolute;
    left: calc(100% + 10px);
    top: 14px;
    width: 20px;
  }

  .padding-icon {
    padding-left: 40px;
  }
}
