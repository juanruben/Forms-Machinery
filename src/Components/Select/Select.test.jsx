import React from 'react';
import { shallow } from 'enzyme';
import Select from './Select';

describe('Select', () => {
    let wrapper;
    let callback;
    beforeEach(() => {
        callback = jest.fn();
        wrapper = shallow(<Select name="test" onChange={callback} />);
    });

    it('should render correctly', () => expect(wrapper.debug()).toMatchSnapshot());

    it('value check', () => {
        wrapper.find('select').simulate('change', { target: { value: 'test' } });
        expect(callback).toHaveBeenCalled();
    });

    it('should render text', () => {
        wrapper.setProps({ value: 'test' });
        expect(wrapper.find('select').prop('value')).toEqual('test');
    });

    it('should render error correctly', () => {
        wrapper.setProps({ errors: 'test' });
        expect(wrapper.find('i')).toHaveLength(1);
        expect(wrapper.find('select').hasClass('border-error')).toBe(true);
    });

    it('should render input', () => {
        expect(wrapper.find('select')).toHaveLength(1);
    });

    it('should render div', () => {
        expect(wrapper.find('div')).toHaveLength(1);
    });
});
