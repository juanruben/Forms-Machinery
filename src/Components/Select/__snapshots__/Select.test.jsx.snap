// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
"<Fragment>
  <div className=\\"input-bordered-container\\">
    <select name=\\"test\\" id=\\"\\" className=\\"select-icafal\\" defaultValue=\\"0\\" onChange={[Function: mockConstructor]}>
      <option value=\\"0\\" disabled={true} />
    </select>
  </div>
</Fragment>"
`;
